Add HLSL shaders and improve tooling

SaschaWillems opened the pull request:
## Description

With the framework now being able to consume shaders in SPIR-V format, this PR will start adding actual HLSL shaders to the repository for at least some samples. I'll be starting with adding HLSL shaders to my samples and maybe other not so complex samples.

### ✅ What this PR currently does:
- Adds a way to specify a set of HLSL shaders for a sample similar to how GLSL shaders can be added to a sample (so they show up in the IDE)
- Adds a new overload for the (hpp)api sample base class that handles toggling between glsl and hlsl
- Adds an optional CMake argument for passing additional arguments to DXC (e.g. to enable a given extension) called `DXC_ADDITIONAL_ARGUMENTS`
- Has some minor fixes for HLSL offline compilation introduced in #990
- Restructures folders for shaders, so samples can have a set of shaders in a glsl sub folder and a hlsl sub folder
  - shaders\sample_name (e.g. ray_tracing_basic)
    - glsl
      - raygen.rgen
      - closesthit.rchit
      - miss.rmiss
    - hlsl
      - raygen.rgen.hlsl
      - closesthit.rchit.hlsl
      - miss.rmiss.hlsl
  - I'm adding .hlsl to the file names as that automatically gives you HLSL syntax highlighting in e.g. MSVC
  - The only downside to this is that we need to exclude these files from the build process, or else MSVC want's to compile them with it's own DXC version for the D3D target
    - To avoid that, I disable this behavior for MSVC, see https://github.com/SaschaWillems/Vulkan-Samples/commit/8aa9573289f65279edd80ac20808f668c6ca0380
- Adds some basic information on the shaders used in this repo incl. links to the HLSL chapters in the Vulkan Guide
- Fixes some GLSL shaders to make mapping to HLSL possible/easier (e.g. HDR sample shaders)
- Adds shaders to some samples where they were missing from the CMake file (and thus wouldn't show up in the IDE)
- Adds a way to switch between GLSL and HLSL shaders via command line using a plugin
  - e.g. `vulkan_samples ray_tracing_basic --shading-language hlsl`
  - The sample base class defaults to glsl, so omitting above argument will try to load glsl shaders

### 📃List of samples that got HLSL shaders added
For a complete list of samples see #1014
This PR adds HLSL shaders (based on their GLSL counterparts) for the following samples:
- calibrated_timestamp
- color_write_enable
- compute_nbody (+hpp variant)
- conditional_rendering
- conservative_rasterization
- debug_utils
- descriptor_buffer_basic
- descriptor_indexing
- dynamic_blending
- dynamic_line_rasterization
- dynamic_primitive_clipping 
- dynamic_rendering
- dynamic_uniform_buffers (+hpp variant)
- fragment_shading_rate
- graphics_pipeline_library 
- hdr (+hpp variant)
- instancing (+hpp variant) 
- logic_op_dynamic_state
- memory_budget
- mesh_shading (+hpp variant)
- open_cl_interop (needs #1021 to work)
- open_gl_interop
- patch_control_points
- portability
- profiles (can't test, see #1008)
- push_descriptors
- ray_tracing_basic
- ray_tracing_extended
- ray_queries
- shader_debugprintf
- separate_image_sampler (+hpp variant)
- sparse_image
- synchronization_2
- terrain_tessellation (+hpp variant)
- texture_compression_basisu
- texture_loading (+hpp variant)
- texture_mipmap_generation (+hpp variant)
- timestamp_queries (+hpp variant)
- vertex_dynamic_state

### ❌ What this won't do
- Add HLSL compilation to the build process (which is not the scope of this PR, see #990, which has been merged)
- Add HLSL shaders for the ARM framework based samples 

### 💬Up for discussion
- ✅ Offline compilation from #990 will result in .spv files generated at build-time. Should we add them to the repo, or exclude them?
  - As per the call on 2024-04-08 we decided to add the compiled .spv files, so people can test/run this without requiring DXC
- How to handle if someone requests hlsl shaders but a sample doesn't have them? Should we just fail or fall back?
- One thing that feels kinda odd now is that we also have a realtime shader selection plugin that was merged from another PR, which defines enums and code that now feels out of place. We should discuss if we want to remove that one. It's not used anywhere.

### 🚧Opens
- Add some more documentation for maintainers on how to get HLSL shaders added to a sample
- Due to the way DXC optimizes vertex inputs, some of the samples display a validation layer performance message for that. It's not critical and nothing that should stop merging us this PR

Fixes #892
Fixes #179

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

SaschaWillems committed:
Use CMake to add (optional) HLSL shaders to project

SaschaWillems committed:
Added HLSL shaders

SaschaWillems committed:
Add python script for HLSL shader compilation
Stop-gap until we have added a more convenient way of compiling HLSL

SaschaWillems committed:
Exclude HLSL shaders form MSVC build
MSVC automatically compiles any HLSL shader using it's built-in DXC version with DX syntax, which obviously fails for VK HLSL shaders

SaschaWillems committed:
Update copyrights

SaschaWillems committed:
Remove .hlsl. spir-v output file name path
Makes it easier to adpot HLSL loading

SaschaWillems committed:
Add basic shader documentation with links to the guide

SaschaWillems committed:
target-env and extensions for RT shaders

asuessenbach commented on samples/extensions/ray_tracing_basic/CMakeLists.txt:
Is it standard to have a common file extension for HLSL files, but not for GLSL files?

SaschaWillems replied:
Yes. Plugins/Addons for GLSL do highlighting based on shader stage extension (.vert, .frag, etc.). Plugins/Addons for HLSL rely on having .hlsl as extension (e.g. MSVC).

SaschaWillems committed:
Merge remote-tracking branch 'upstream/main' into hlsl_shaders_extension_samples

# Conflicts:
#	bldsys/cmake/sample_helper.cmake

SaschaWillems committed:
Fix merge errors

SaschaWillems committed:
Fix typo in shader extension detection

SaschaWillems committed:
Adjust shader folder name

SaschaWillems committed:
Adjust HLSL SPIR-V output file name
Remove the .hlsl. part to make loading in the framework easier

SaschaWillems committed:
Add new overload for loading shadders

Separates folder and shader file name
Will be the base for shader language selection

SaschaWillems committed:
Add shading language property to vulkan sample base class

SaschaWillems committed:
Add new plugin for shading language selection via command line arguments

SaschaWillems committed:
Remove python HLSL compilation script. No longer required with HLSL getting compiled as part of CMake

SaschaWillems committed:
Move shading_language member to proteced, so samples can access it

SaschaWillems committed:
Api samples now load shaders bases on shading language selection

SaschaWillems committed:
Trying to get past clang-format

SaschaWillems committed:
Add HLSL shaders for instancing sample

SaschaWillems committed:
Added HLSL shaders for separate image sampler example

SaschaWillems committed:
Add CMake option for additional DXC arguments

SaschaWillems committed:
Add HLSL shaders for shader debugprintf sample

SaschaWillems committed:
Added HLSL shaders to texture loading sample

SaschaWillems committed:
Added HLSL shaders for dynamic uniform buffer sample

SaschaWillems committed:
Added HLSL shaders to texture mipmap generation sample

SaschaWillems committed:
Add HLSL shaders for conditional rendering sample

SaschaWillems committed:
Added shader loading based on selected shading language to hpp sample class

SaschaWillems committed:
Added HLSL support to HPP dynamic uniform buffer sample

SaschaWillems committed:
Added HLSL shaders to hpp instancing sample

SaschaWillems committed:
Added HLSL shaders to hpp separate image sampler sample

SaschaWillems committed:
Added HLSL shaders to hpp texture loading sample

SaschaWillems committed:
Added HLSL shaders to hpp texture mipmap generation sample

SaschaWillems committed:
Add HLSL shaders for terrain tessellation sample

SaschaWillems committed:
Add HLSL shaders to hpp terrain tessellation sample

SaschaWillems committed:
Added HLSL support to memory budget sample

SaschaWillems committed:
Added HLSL support to OpenGL interop sample

SaschaWillems committed:
Use enum for shading language selection

SaschaWillems committed:
Added HLSL shaders to texture compression basis universal sample

SaschaWillems committed:
Trying to make clang-format happy

SaschaWillems committed:
Trying to make clang-format happy

SaschaWillems committed:
Added HLSL shaders to push descriptor sample

SaschaWillems committed:
Fix HLSL shader profile for mesh shaders

SaschaWillems committed:
Adjust DXC target environment for mesh shaders

SaschaWillems committed:
Add HLSL support to mesh shader sample (including hpp variant)

SaschaWillems committed:
Add HLSL support to hpp mesh shading sample

SaschaWillems committed:
Added HLSL shaders to conservative rasterization sample

SaschaWillems committed:
Added HLSL shaders to compute nbody example (incl. hpp variant)

SaschaWillems committed:
Added HLSL suppor to the HDR sample (including HPP variant)
Needed to change a few things with the GLSL shaders too in order to match HLSL

SaschaWillems committed:
Add missing pipeline stage

SaschaWillems committed:
Added HLSL support to timestamp query samples (incl. hpp variant)
Those samples are using the shaders from the HDR sample

SaschaWillems committed:
Added HLSL shaders for descriptor buffer basic sample

SaschaWillems committed:
Add HLSL shaders to profile sample
Note: Can't test right now as sample is broken on main

SaschaWillems committed:
Fix copyrights

SaschaWillems committed:
Updated copyrights
To silence CI

SaschaWillems committed:
Added HLSL shaders to dynamic blending sample

SaschaWillems committed:
Added HLSL shaders to dynamic rendering sample

SaschaWillems committed:
Remove python script, no longer required with offline compilation at CMake level

SaschaWillems committed:
Add HLSL shaders to vertex dynamic state sample

SaschaWillems committed:
Added HLSL shaders to synchronization_2 sample

SaschaWillems committed:
Add HLSL shaders to logic op dynamic state sample

asuessenbach commented on app/plugins/shading_language_selection/shading_language_selection.cpp:
```diff
@@ -0,0 +1,63 @@
+		{
+			LOGI("Shading language selection: GLSL");
+			vkb::VulkanSample<vkb::BindingType::C>::set_shading_language(vkb::ShadingLanguage::GLSL);
+			vkb::VulkanSample<vkb::BindingType::Cpp>::set_shading_language(vkb::ShadingLanguage::GLSL);
```
Instead of setting this value twice, once for C- and once for C++-binding, you could move this functionality into the vkb::Application class and call it just once.
Or, what I would prefer, add this to the vkb::ApplicationOptions and pass it in via Application::prepare (similar to what I've asked for in #829).

SaschaWillems replied:
I tried to but wasn't able. Probably due to my lack of C++ skills. Will try again.

SaschaWillems replied:
Moved to the application base class.

asuessenbach commented on bldsys/cmake/sample_helper.cmake:
```diff
@@ -22,10 +23,10 @@ set(SCRIPT_DIR ${CMAKE_CURRENT_LIST_DIR})
 function(add_sample)
     set(options)  
-    set(oneValueArgs ID CATEGORY AUTHOR NAME DESCRIPTION)
+    set(oneValueArgs ID CATEGORY AUTHOR NAME DESCRIPTION DXC_ADDITIONAL_ARGUMENTS)
```
Is `DXC_ADDITIONAL_ARGUMENTS` a `oneValueArg`? The name looks like a `multiValueArg`.

SaschaWillems replied:
Yeah, it's a oneValueArg. I might rename or change depending on where I'll be needing this.

asuessenbach commented on bldsys/cmake/sample_helper.cmake:
```diff
@@ -66,14 +68,20 @@ function(add_sample_with_tags)
+        foreach(SHADER_FILE_GLSL ${SHADER_FILES_GLSL})
+            list(APPEND SHADERS_GLSL "${PROJECT_SOURCE_DIR}/shaders/${SHADER_FILE_GLSL}")
+        endforeach()        
+    endif()
```
Why do you change this?

SaschaWillems replied:
Can you elaborate? Change in regards to what? Main? If so, probably due to some merge conflict.

asuessenbach replied:
Well, according to the diff-tool, you changed from
```
    foreach(SHADER_FILE_GLSL ${TARGET_SHADER_FILES_GLSL})
        list(APPEND SHADERS_GLSL "${PROJECT_SOURCE_DIR}/shaders/${SHADER_FILE_GLSL}")
    endforeach()
```
back to
```
    if (TARGET_SHADER_FILES_GLSL)
        list(APPEND SHADER_FILES_GLSL ${TARGET_SHADER_FILES_GLSL})
        foreach(SHADER_FILE_GLSL ${SHADER_FILES_GLSL})
            list(APPEND SHADERS_GLSL "${PROJECT_SOURCE_DIR}/shaders/${SHADER_FILE_GLSL}")
        endforeach()        
    endif()
```

asuessenbach commented on bldsys/cmake/sample_helper.cmake:
```diff
@@ -66,14 +68,20 @@ function(add_sample_with_tags)
+        foreach(SHADER_FILE_HLSL ${SHADER_FILES_HLSL})
+            list(APPEND SHADERS_HLSL "${PROJECT_SOURCE_DIR}/shaders/${SHADER_FILE_HLSL}")
+        endforeach()        
+    endif()
```
As above: why do you change this?

SaschaWillems replied:
Reverted.

asuessenbach commented on bldsys/cmake/sample_helper.cmake:
```diff
@@ -142,14 +151,14 @@ function(add_project)
+    if (SHADERS_HLSL)
+        source_group("\\Shaders\\hlsl" FILES ${SHADERS_HLSL})
+        # Disable automatic compilation of HLSL shaders for MSVC
+        set_source_files_properties(SOURCE ${SHADERS_HLSL} PROPERTIES VS_SETTINGS "ExcludedFromBuild=true")        
```
I see, it works, but shouldn't `TARGET_SHADERS_[GLSL|HLSL]` be used here?

SaschaWillems replied:
Yeah, both work, but the TARGET_ is more correct. Will change.

asuessenbach commented on framework/vulkan_sample.h:
```diff
@@ -334,6 +340,9 @@ class VulkanSample : public vkb::Application
 	static void set_viewport_and_scissor(CommandBufferType const &command_buffer, Extent2DType const &extent);
 
+	/** @brief Used to select between different shader languages, static so it can be changed from a plugin */
+	inline static vkb::ShadingLanguage shading_language{vkb::ShadingLanguage::GLSL};
```
In order to enforce data encapsulation, I would prefer to have the member variables be private.

SaschaWillems replied:
Makes sense, will change.

SaschaWillems replied:
After moving this to the application base class this is now private and accessed with a getter.

asuessenbach commented on samples/extensions/shader_debugprintf/CMakeLists.txt:
```diff
@@ -23,8 +23,13 @@ add_sample(
     ID ${FOLDER_NAME}
     CATEGORY ${CATEGORY_NAME}
     AUTHOR "Sascha Willems"
+    FUUU "fuuu"
```
??

asuessenbach reviewed (commented):
What a nice little PR!!

I don't know much about HLSL, so I didn't look at those files.
All the rest looks pretty good to me, just a couple of mostly minor issues.

And:
> How to handle if someone requests hlsl shaders but a sample doesn't have them? Should we just fail or fall back?

I think, failure on loading a shader is an acceptable behaviour for this framework. No matter if it's glsl or hlsl.

SaschaWillems committed:
Remove accidentally commited line

SaschaWillems committed:
Added HLSL shaders for sparse image sample

SaschaWillems committed:
Added HLSL shaders to patch control points sample

SaschaWillems committed:
Move shading language selection to application base class

SaschaWillems committed:
Trying to fix CI

SaschaWillems committed:
Adjust sample cmake helper based on reivew

SaschaWillems committed:
Revert some changes

SaschaWillems committed:
Enable mesh shader SPIR-V extension for DXC

SaschaWillems committed:
Added HLSL shaders to ray queries sample

SaschaWillems committed:
Added HLSL shaders to ray tracing extended sample
Mostly working, still need some fixing

SaschaWillems committed:
Update copyrights

SaschaWillems committed:
Fxied HLSL shaders, now visually matching GLSL ones

SaschaWillems committed:
Rename folder to be in line with sample name

SaschaWillems committed:
Merge remote-tracking branch 'upstream/main' into hlsl_shaders_extension_samples

SaschaWillems committed:
Added HLSL shaders to color write enable sample

SaschaWillems committed:
Added HLSL shaders to debug utils sample

SaschaWillems committed:
Added HLSL shaders to portability sample
Removed unnecessary shader loading function

SaschaWillems committed:
Trying to fix CI...

SaschaWillems committed:
Added HLSL shaders to descriptor indexing sample

SaschaWillems committed:
Added HLSL shaders to dynamic line rasterization sample

SaschaWillems committed:
Semantic names

SaschaWillems committed:
Remove HLSL warning

SaschaWillems committed:
Added HLSL shaders to dynamic primitive clipping sample

SaschaWillems commented:
Quick note: While adding HLSL shaders I noticed that several samples are broken on main or at least display validation errors. So when reviewing, most errors that show up are probably also visible on main. I didn't want to fix it in order not to pollute this PR.

SaschaWillems committed:
Added SPIR-V files compiled from HLSL shaders

SaschaWillems committed:
Added HLSL shaders to fragment shading rate sample

SaschaWillems committed:
Added SPIR-V files

SaschaWillems committed:
Added HLSL shaders to graphics pipeline library example

SaschaWillems committed:
Added HLSL shaders to OpenCL interoperability sample

SaschaWillems committed:
Merge remote-tracking branch 'upstream/main' into hlsl_shaders_extension_samples

JoseEmilio-ARM commented:
> Quick note: While adding HLSL shaders I noticed that several samples are broken on main or at least display validation errors. So when reviewing, most errors that show up are probably also visible on main. I didn't want to fix it in order not to pollute this PR.

Tested most samples and found no issues, sorry for the delay. I did notice that Performance/WaitIdle does not launch on Android. I thought it was unrelated, but the same sample opens fine in #1051 , maybe it's an issue from a commit in between the two?

SaschaWillems commented:
> I did notice that Performance/WaitIdle does not launch on Android. I thought it was unrelated

That is indeed unrelated to this PR. That sample was broken on master for quite some time and was fixed by a different PR that I haven't merged. So it shouldn't matter for this PR :)
